File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.js
class Solution {
    isSubtree(root, subRoot) {
        if (subRoot === null) return true;
        if (root === null) return false;

        if (this.isSame(root, subRoot)) {
            return true;
        }

        return (
            this.isSubtree(root.left, subRoot) ||
            this.isSubtree(root.right, subRoot)
        );
    }

    isSame(p, q) {
        if (p === null && q === null) return true;
        if (p === null || q === null) return false;
        if (p.val !== q.val) return false;

        return (
            this.isSame(p.left, q.left) &&
            this.isSame(p.right, q.right)
        );
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node, min_val, max_val):
            if node is None:
                return True

            if node.val <= min_val or node.val >= max_val:
                return False

            return (
                dfs(node.left, min_val, node.val) and
                dfs(node.right, node.val, max_val)
            )

        return dfs(root, -float('inf'), float('inf'))